Snowflake.Data/Client/SnowflakeDbConnection.cs

```diff
@@ -86,17 +99,53 @@ public override string DataSource
             }
         }
 
-        public override string ServerVersion => _connectionState == ConnectionState.Open ? SfSession.serverVersion : "";
+        public override string ServerVersion => IsOpen() ? SfSession.serverVersion : "";
 
         public override ConnectionState State => _connectionState;
 
+        private object? GetTransactionId()
+        {
+            if (!IsOpen())
+            {
+                return DBNull.Value;
+            }
+            using (IDbCommand command = CreateCommand())
+            {
+                command.CommandText = "SELECT CURRENT_TRANSACTION()";
+                return command.ExecuteScalar(); 
+            }
+        }
+
+        public bool HasTransactionInProgress() => GetTransactionId() != DBNull.Value;
+        
+        private bool TerminateTransaction(object? transactionInProgressId)
+        {
+            if (transactionInProgressId == DBNull.Value)
+                return false;
+            try
+            {
+                using (IDbCommand command = CreateCommand())
+                {
+                    command.CommandText = "ROLLBACK";
+                    command.ExecuteNonQuery(); 
+                    logger.Warn("Rolled back transaction: " + transactionInProgressId + " in session: " + SfSession.sessionId);

```

Comment:
why is it warn? someone ordered transaction termination on purpose

Reply:
It's a private operation used within closing a connection. If it was a publicly available interace I would agree with you but right now this is just to point out for a user that he's returning connection to the pool without calling ROLLBACK/COMMIT on his own - that's why it's a WARN.
Maybe I change a name of this function like: TerminateTransactionForDirtyConnectionClose?

Reply:
I changed messages and TerminateTransaction name, check if this is clear now